### PR TITLE
Ch04 Refactor

### PR DESCRIPTION
**Pull request by dcrouch26**
1. Cleans Chapter 4 code
2. Adds run_cell extension to load individual cells
3. Fixes tabulation in aips
4. Adds Jupyter flag allowing for local connections

### CODE REVIEW

File: docker/data-science/Dockerfile [hunk 1]
```diff
@@ -52,4 +52,5 @@ RUN find . -name \*.ipynb -print0|xargs -0 jupyter-trust -y
 
 RUN jupyter labextension disable "@jupyterlab/apputils-extension:announcements"
 CMD start-notebook.sh --ServerApp.token='test' --ServerApp.password='' --ServerApp.ip=0.0.0.0 --no-browser \
-  --LabApp.default_url='/lab/tree/welcome.ipynb' --NotebookApp.token='token' --LabApp.token='test'
+  --LabApp.default_url='/lab/tree/welcome.ipynb' --NotebookApp.token='token' --LabApp.token='test' \
+  --NotebookApp.allow_origin='*' --NotebookApp.ip='0.0.0.0'
```
Comment (alexott) [inline docker/data-science/Dockerfile:56]: This really should be fixed in #139

File: docker/data-science/notebooks/aips.py [hunk 2]
```diff
@@ -23,44 +23,44 @@
...
 
-  try:
-    if (get_engine().health_check()):
-      print ("Solr is up and responding.")
-      print ("Zookeeper is up and responding.\n")
-      print ("All Systems are ready. Happy Searching!")
-  except:
-      print ("Error! One or more containers are not responding.\nPlease follow the instructions in Appendix A.")
+    try:
+        if (get_engine().health_check()):
```
Comment (alexott) [inline docker/data-science/notebooks/aips.py:29]: do we need to have `if` here if the function throws an exception anyway?

File: docker/data-science/notebooks/aips.py [hunk 3]
```diff
@@ -285,9 +285,39 @@ def img_path_for_upc(upc):
...
+        x = re.search(results_template_syntax, file_content, flags=re.S)
+        results_template = x.group(1)
+
+        separator_template_syntax = "<!-- BEGIN_TEMPLATE: SEPARATOR -->(.*)<!-- END_TEMPLATE: SEPARATOR[^>]*-->"
+        x = re.search(separator_template_syntax, file_content, flags=re.S)
+        separator_template = x.group(1)
+
+        rendered = header_template.replace("${QUERY}", query)
+        for result in documents:
+            rendered += results_template.replace("${NAME}", result['name'] if 'name' in result else "UNKNOWN") \
```
Comment (alexott) [inline docker/data-science/notebooks/aips.py:310]:
```
result['name'] if 'name' in result else "UNKNOWN"
```
is easier to replace with
```
result.get('name', "UNKNOWN")
```

same for other lines

File: docker/data-science/notebooks/aips.py [hunk 4]
```diff
@@ -285,9 +285,39 @@ def img_path_for_upc(upc):
...
+
+        results_template_syntax = "<!-- BEGIN_TEMPLATE: SEARCH_RESULTS -->(.*)<!-- END_TEMPLATE: SEARCH_RESULTS[^>]*-->"
+        x = re.search(results_template_syntax, file_content, flags=re.S)
+        results_template = x.group(1)
+
+        separator_template_syntax = "<!-- BEGIN_TEMPLATE: SEPARATOR -->(.*)<!-- END_TEMPLATE: SEPARATOR[^>]*-->"
+        x = re.search(separator_template_syntax, file_content, flags=re.S)
+        separator_template = x.group(1)
+
+        rendered = header_template.replace("${QUERY}", query)
```
Comment (alexott) [inline docker/data-science/notebooks/aips.py:308]: We can also use https://docs.python.org/3.4/library/string.html#template-strings to replace templates

Comment (dcrouch26) [reply docker/data-science/notebooks/aips.py:308]: I'll do a swift refactor on this templating logic when the major chapter refactor is finished, these changes were marked when I fixed tabulation

File: docker/data-science/notebooks/aips.py (hunk 2 above)
Comment (dcrouch26) [reply docker/data-science/notebooks/aips.py:29]: There are instances where health check doesn't throw an exception but the healthcheck fails. I'm seeing now that the else case doesn't return the failure message. Will fix that

**Comment by dcrouch26** [comment]
Will be closing this PR and creating a temporary parallel

**Pull request closed without merging**